### PR TITLE
Add support for backup codes

### PR DESCRIPTION
**Pull request by sbc100**
Backup codes also generated randomly and stored one-way-hashed
in the database using Devise::Encryptor.  They default to 8
decimal digits rather than 6 since they are longer lived than
direct OTP codes.

### CODE REVIEW

**Comment by sbc100** [comment]
This case is not quite ready for merging but I wanted to get some eyes on it early.

**Comment by sbc100** [comment]
@monfresh what do you think?

File: app/controllers/devise/two_factor_authentication_controller.rb [hunk 1]
```diff
@@ -2,16 +2,50 @@ class Devise::TwoFactorAuthenticationController < DeviseController
...
+                    resource.authenticate_totp(params[:totp_code])
+                  end
 
-    if resource.authenticate_otp(params[:code])
+    if auth_status
       after_two_factor_success_for(resource)
     else
-      after_two_factor_fail_for(resource)
+      after_two_factor_fail_for(resource, type)
     end
```
Comment (monfresh) [inline app/controllers/devise/two_factor_authentication_controller.rb]: This method is doing too many things. It violates the Single Responsibility Principle. If you ran Rubocop against this method, it would raise several offenses. One way to find out if a class or method is doing too much is to describe it in plain English. In this case, you might say that this class verifies 3 different methods of two factor authentication. Sometimes it verifies a direct OTP, sometimes it verifies a TOTP, and sometimes it verifies a backup code. That's 3 different things this class is responsible for.

I would recommend moving each method of verification into its own controller.

Comment (monfresh) [inline app/controllers/devise/two_factor_authentication_controller.rb]: What do you think about splitting this PR into two? Or for now, removing the controller stuff, and only implementing the code creation and validation, and leaving the controller implementation to the user. This would allow people to take advantage of the backup code right away while we decide on the best way to implement the controllers, or if we event want to. Maybe some users might only need a subset of the 3 types, and might want to implement it their own way.

@sbc100 @Houdini What do you think?

File: lib/two_factor_authentication/models/two_factor_authenticatable.rb [hunk 2]
```diff
@@ -87,6 +103,19 @@ def generate_totp_secret
           ROTP::Base32.random_base32
         end
 
+        def create_backup_codes(options = {})
```
Comment (monfresh) [inline lib/two_factor_authentication/models/two_factor_authenticatable.rb]: Wanna document the usage of this method in the README? It's only being called in the tests in this PR.

Comment (monfresh) [inline lib/two_factor_authentication/models/two_factor_authenticatable.rb]: What do you think about putting this in its own module, like the `devise-two-factor` gem? https://github.com/tinfoil/devise-two-factor/blob/master/lib/devise_two_factor/models/two_factor_backupable.rb

File: lib/two_factor_authentication/models/two_factor_authenticatable.rb [hunk 3]
```diff
@@ -87,6 +103,19 @@ def generate_totp_secret
...
+          count = options[:count] || self.class.backup_code_count || 10
+          digits = options[:length] || self.class.backup_code_length || 8
+          backup_codes = (1..count).map do
+            random_base10(digits)
```
Comment (monfresh) [inline lib/two_factor_authentication/models/two_factor_authenticatable.rb]: Any reason to use numbers only for codes as opposed to alphanumeric characters, like the devise-two-factor gem? GitHub and Apple also use alphanumeric codes.

Comment (monfresh) [inline lib/two_factor_authentication/models/two_factor_authenticatable.rb]: - **AOL**: numeric only, 16 characters in groups of 4, separated by a dash
- **Dropbox**: alphanumeric, 16 characters in groups of 4, separated by a space
- **GitHub**: alphanumeric, 10 characters in groups of 5, separated by a dash
- **Heroku**: alphanumeric, 16 characters, all lowercase
- **Hover**: alphanumeric, 16 characters in groups of 4, separated by a space
- **IFTTT**: numeric only, 16 characters in groups of 4, separated by a space
- **Mailchimp**: alphanumeric, 16 characters, all caps
- **Stripe**: alpha only, 20 characters, 5 groups of 4, separated by dash

**Comment by Houdini** [comment]
I hope @monfresh could finish his review and merge it.
Let's give him several days more.

**Comment by ziaulrehman40** [comment]
I need some features from this gem, but also backup codes is needed, so unless this is merged, I am forced to either write extension for the other gem which I am considering or wait for this PR to get merged.

**Comment by monfresh** [comment]
This PR needs more work before it gets merged. I'm not sure when I'll have time to work on it. Maybe next week. If someone else has the time to clean it up before me, I'd be happy to review it.

**Comment by sbc100** [comment]
Closing since this is two years old now..
